### DISCUSSION TITLE
number_density(electron) returns 0 when loading from file?

### DISCUSSION THREAD

**Discussion by fcour** [epochpic/epoch#734] [Q&A] [answered]
Dear all,
I'm currently building an input deck for a 3D simulation where electron density is generated from a binary file. However, when generating the corresponding ion species to get electrical neutrality, the commands density(electron) or number_density(electron) return 0 [example below]. In contrast, if I specify the electron density using mathematical expression, everything goes well. 
Any idea? Thanks in advance!

```
begin:species
    name = electron
    mass = 1.0
    charge = -1.0
    npart_per_cell = part
    density = 'InputDensity.bin'  
    temp = 240000
end:species

begin:species
  name = ion1
  charge = 1.0
  mass = neutral_mass-1
  npart_per_cell = part
  immobile = T
  density = number_density(electron)
  temp_ev = 1e-3
end:species

```

**Comment by Status-Mirror** [comment]
I'm not sure the binary file loading is compatable with the EPOCH maths parser. In this example, you could just set the `ion1` density to also be `InputDensity.bin` though. Are you able to load the `electron` species okay?

Cheers,
Stuart

**Reply by fcour** [reply]
yes, setting ion1 density to "InputDensity.bin" works. However, -if I've well understood- the positions of the particles are not the same between electrons and ions implying that the E-fields are not zero. 
Besides, loading electron species is okay.
thanks for your feedback!

**Reply by Status-Mirror** [reply, marked as answer]
That's correct, but you'd have this problem of non-overlapping particle positions even if you used the maths parser. Macro-particle positions in cells are always random. The only way you could override this behaviour is to manually specify the starting position of each particle, as in the [load particles from file demo](https://epochpic.github.io/quickstart/basic_examples/particle_load.html#load-particles-from-file).

Edit: The E-fields are assumed 0 at the simulation start by the way. If you have a chage imbalance, the code will assume there is an immobile charge distribution present to initially neutralise each cell. This will cause issues if all your particle move away - this immobile charge may affect their trajectories. In practice, the energies of plasma particles mean that the small field correction caused by the initial immobile charge is usually too small to care about.

**Reply by fcour** [reply]
is the non-overlapping position also true when using ion=number_density(electron) ?
thanks for the suggestion with 0-initial field.

**Reply by Status-Mirror** [reply]
Yes - "ion = number_density(electron)" will set the density values of each cell, but macro-particles will still be loaded at random positions within the cell. There will not be a perfect overlap between electrons and ions, even if both species have the same number densities in the same cells.

The only way you can force exact particle positions is to specify them manually.

**Reply by fcour** [reply]
thanks!
